controllers/utils/dns.go

```diff
@@ -64,8 +64,30 @@ func Dig(fqdn string, edgeDNSServers ...DNSServer) (ips []string, err error) {
 	if err != nil {
 		return nil, fmt.Errorf("dig error: %s", err)
 	}
+	aRecords := make([]*dns.A, 0)
+	cnameRecords := make([]*dns.CNAME, 0)
 	for _, a := range ack.Answer {
-		ips = append(ips, a.(*dns.A).A.String())
+		switch v := a.(type) {
+		case *dns.A:
+			ips = append(ips, v.A.String())
+			aRecords = append(aRecords, v)
+		case *dns.CNAME:
+			cnameRecords = append(cnameRecords, v)
+		}
+	outer:
+		for _, cname := range cnameRecords {
+			for _, a := range aRecords {
+				if cname.Target == a.A.String() {
```

Comment:
If you want to avoid duplicates I suggest using an additional map to keep track of the IP addresses found. Something along these lines:
```
	seen := make(map[string]struct{}) 
	for _, a := range ack.Answer {
		ip := a.(*dns.A).A.String()
		if _, exists := seen[ip]; !exists {
			seen[ip] = struct{}{} // Mark the IP as seen
			ips = append(ips, ip) // Add the IP to the slice
		}
        }
```
I think it will be easier to follow than using loop labels

Reply:
> What is the purpose of this check?

This is to ensure that in the answer returned by the DNS server, every CNAME record has a corresponding A record.  It MIGHT be possible for a non-recursive resolver or misconfigured resolver to only return the CNAME.

After processing both CNAMEs and A records, if we have any CNAME records referencing A records that were not present in the reply from the server, run another request for the dereferenced A record.

This also made it easier to unit test in the mock, as the mock server returns ONLY matching records -- it returns just the CNAME in the first request, and then we re-request the A record.

I can rewrite it with a tracking map if it's easier to read -- I'm still adapting to golang idioms from python idioms.

Reply:
I also just realized I had this inside the loop checking answers, instead of outside of it. whoops!

Reply:
Can you give an example, I still didn't fully understand.
You have: `if c.Target == a.A.String() {`
If c is `foo.example.com CNAME bar.example.com`. And a is `foobar.example.com A 1.2.3.4`. Then you are comparing `bar.example.com` with `1.2.3.4` which will never be true.

[This post](https://serverfault.com/questions/1049316/is-it-unusual-for-a-dns-response-to-contain-both-a-records-and-cname-records) says a response cannot contain both a CNAME and an A record. Also, I believe it is invalid to have two CNAMEs in a reply.
If that holds we can do:
```
	aRecords := make([]*dns.A, 0)
	cnameRecords := make([]*dns.CNAME, 0)
	for _, a := range ack.Answer {
		switch v := a.(type) {
		case *dns.A:
			ips = append(ips, v.A.String())
			aRecords = append(aRecords, v)
		case *dns.CNAME:
			cnameRecords = append(cnameRecords, v)
		}
	}

        if len(cnameRecords) > 0 and len(aRecords) > 0 {
           // error
        }
        if len(aRecords) > 0 {
           // return A records
        }
        if len(cnameRecords) > 1 {
          // error
        }

        // resolve cname records
```

Reply:
A DNS resolver can send a CNAME record pointing to another name, and one or more A records for the name targeted by the CNAME in the same response.  The code is intended to check for targets of CNAMEs that do not have an A record included in the same response.

However, with further thought, I don't think this code works correctly in a situation where the following RRs are present in an answer:

CNAME Alpha targets Beta
CNAME Beta targets Charlie
Charlie has A record 1.2.3.4

In this case, we have a complete chain ending in at least one A record, but the code, as I have it now, would count A CNAME B as "unresolved" and send another request asking for an A record for Beta and would get Charlie, and proceed normally, with an extra round trip.

I'm going to take a break for US holidays, but will come back to this the middle of next week to re-examine.

Reply:
See below, an example of the DNS query/response for a service running in one of our clusters, that we handle via hostNetwork ingress-nginx.  The DNS server at 172.0.0.1 is an Infoblox NIOS server.

```pcap
10:46:16.713619 00:00:00:00:00:00 > 00:00:00:00:00:00, ethertype IPv4 (0x0800), length 91: (tos 0x0, ttl 64, id 62866, offset 0, flags [DF], proto UDP (17), length 77)
    172.23.131.235.58932 > 172.0.0.1.53: [bad udp cksum 0x3b14 -> 0xa7b9!] 1030+ A? my-application-amer.example.com. (49)
10:46:16.713668 00:00:00:00:00:00 > 00:00:00:00:00:00, ethertype IPv4 (0x0800), length 91: (tos 0x0, ttl 64, id 62867, offset 0, flags [DF], proto UDP (17), length 77)
    172.23.131.235.58932 > 172.0.0.1.53: [bad udp cksum 0x3b14 -> 0xddb5!] 45833+ AAAA? my-application-amer.example.com. (49)
10:46:16.714579 00:00:00:00:00:00 > 00:00:00:00:00:00, ethertype IPv4 (0x0800), length 252: (tos 0x0, ttl 57, id 10334, offset 0, flags [none], proto UDP (17), length 238)
    172.0.0.1.53 > 172.23.131.235.58932: [udp sum ok] 1030* q: A? my-application-amer.example.com. 5/0/0 my-application-amer.example.com. CNAME ingress.amer-cluster.ks.example.com., ingress.amer-cluster.ks.example.com. CNAME ingress-nginx-controller.kube-system.svc.amer-cluster.ks., ingress-nginx-controller.kube-system.svc.amer-cluster.ks. A 172.17.7.12, ingress-nginx-controller.kube-system.svc.amer-cluster.ks. A 172.17.2.166, ingress-nginx-controller.kube-system.svc.amer-cluster.ks. A 172.17.3.51 (210)
10:46:16.714608 00:00:00:00:00:00 > 00:00:00:00:00:00, ethertype IPv4 (0x0800), length 276: (tos 0x0, ttl 57, id 10335, offset 0, flags [none], proto UDP (17), length 262)
    172.0.0.1.53 > 172.23.131.235.58932: [udp sum ok] 45833* q: AAAA? my-application-amer.example.com. 2/1/0 my-application-amer.example.com. CNAME ingress.amer-cluster.ks.example.com., ingress.amer-cluster.ks.example.com. CNAME ingress-nginx-controller.kube-system.svc.amer-cluster.ks. ns: amer-cluster.ks. SOA ns.dns.amer-cluster.ks. hostmaster.amer-cluster.ks. 1733417154 7200 1800 86400 30 (234)

```

In this case we have a CNAME pointing at a CNAME, pointing at multiple A records for the ingress controllers, all returned in a single DNS packet/response

Reply:
I think this is all covered in the latest changeset, please feel free to start a new conversation if there are additional questions.